brave/brave-core
# Add support for referral promos

Fixes brave/brave-browser#287

Also relies on https://github.com/brave/brave-browser/pull/1244

## Submitter Checklist:


## Automated Test Plan
`npm run test -- brave_unit_tests --filter=BraveReferralsNetworkDelegateHelperTest.*`

## Test Plan:

- Install a version of Brave with an associated referral code in its filename (on Mac, this requires running the .pkg installer not the .dmg)
  - Verify that after installation but before running Brave for the first time, `promoCode` file is in `user-data-directory` and contains the appropriate code. Also verify that this works as expected when the installer is run from a folder containing spaces.
- Run installed Brave
  - Verify that promo-appropriate URL (if any) opens in a new browser tab.
  - Verify that when visiting a referral domain, we send the promo-appropriate request headers.
  - Verify that we no longer send referral code to update server after 90 days.
  - Verify that `promoCode` file is deleted after promo code is read and stored in `local_state` preferences.
  - Verify that service checks for referral finalization 30 days after referral initialization.
  - Verify that service retries checking for referral finalization up to 30 times (with 24-hour pauses between attempts) if it doesn't receive "true" on its first try.
  - Verify that stats.brave.com includes referral code when appropriate.

## Reviewer Checklist:
